### DISCUSSION TITLE
Multiple Eureka discovery services

### DISCUSSION THREAD

**Discussion by famod** [smallrye/smallrye-stork#952] [Q&A]
Hi everyone,

I'm planning on using Stork in a Keycloak extension which shall load balance (and kind of failover) over a couple of Spring services for user federation purposes.
Those Spring services are registered in Spring Cloud Discovery (= Eureka).
Judging from the docs, connecting Stork to an Eureka instance shouldn't be much of a problem (even though we might need to do it purely programmatically, given that you cannot use ArC and friends in a Keycloak extension, sadly).

The thing is that, for high availablity purposes, there are multiple discovery instances and not just a single one.
How can I utilize all those instances? Is the "Composite Service Discovery" the right way to go, meaning composing multiple Eureka discovery services?

Cheers!

**Comment by cescoffier** [comment]
Yes, you can use the composite service discovery to do this. If you configure it with your _n_ eureka instances, It will aggregate the services form the the different instances. Then, in the selection, you would need to make sure the selection is _fair_. 

Another solution is to create your own discovery support (you can implement an SPI - no need for Arc/CDI), which connects to your Eureka and returns the service instances but removes duplicates, so the selection is _fair_.

**Reply by famod** [reply]
Thanks @cescoffier!

Question about the duplicates: Say I have 2 Eureka instances and in each of those the same 4 instances of a certain service are regsitered.
In that case I wouldn't need to worry about the duplcates (and could just use the composite discovery) because each service is returned twice so selection would still be fair, right?

Also, with the composite discovery, I assume that if one Eureka instance is down or throws errors, it simply uses the other instance and doesn't complain, yes?

**Reply by cescoffier** [reply]
> In that case I wouldn't need to worry about the duplcates (and could just use the composite discovery) because each service is returned twice so selection would still be fair, right?

If everything is registered everywhere, no problem. When you have partitions, it's more problematic. 

> Also, with the composite discovery, I assume that if one Eureka instance is down or throws errors, it simply uses the other instance and doesn't complain, yes?

Not sure, @aureamunoz do you know?

**Reply by aureamunoz** [reply]
Hi, excuse me for the delay to respond. If Eureka instance doesn't respond, Stork will complain, you would get an error like:
```
Failed to fetch service instances
io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8661
```

**Reply by famod** [reply]
Thanks @aureamunoz!
That error makes sense to me if using a single Eureka Service Discovery but is this really also happening when using _two_ (or more) Eureka instances via Composite Service Discovery and only one of it is unavailable? (or in other words: at least one of the composed Eureka instances _is_ available)
That would be a bit unfortunate in terms of robustness.

**Reply by aureamunoz** [reply]
Stork doesn't work as a standalone service discovery cluster. Instead, it processes composite configurations—meaning it handles multiple configurations that may use different service discovery implementation. For each specific configuration, Stork delegates the service discovery task to the appropriate service discovery provider. If one of these providers doesn't respond or fails, it affects Stork's ability to resolve that specific configuration, but it doesn't mean Stork itself is faulty; it relies on the performance of the service discovery systems it’s configured to work with.

**Reply by famod** [reply]
I see and on second thought it makes sense given that Composite Service Discovery is really about _composing different_ discovery services that may each return very different services/partitions.
So "Discovery Failover" is not something Composite Service Discovery is aiming to solve.

I think I'll have to implement my own discovery then but I do think it would make sense to have something in Stork OOTB.
Btw, AFAICS, in Spring Boot you have exactly that feature: configure multiple eureka discovery services and on error it will "fail over" to the next available one.

**Reply by aureamunoz** [reply]
Yes, makes sense. I will create an issue and consider it. Moreover, IMO It makes sense it will be implemented in the Eureka service discovery provider, not in the composite side. Thank you for the idea.

**Reply by famod** [reply]
Thanks!

I was thinking about some more generic "wrapper" (similar to composite) so that you are not limited to eureka.

**Reply by aureamunoz** [reply]
Yes, yes, definitely. We are thinking indeed to improve the error management in the composite side too so it can do the work
